# PlasBin-flow: A flow-based MILP algorithm for plasmid contigs binning

## Overview

The identification of plasmids and plasmid genes from sequencing data is an important question regarding antimicrobial resistance spread and other One-Health issues. PasBin-flow is a plasmid bining tool that clusters contigs from a short-read assembly into groups of contigs expected to originate from the same plasmid (called *plasmid contigs bins*). 

PlasBin-flow is based on a Mixed-Integer Linear Programming (MILP) formulation for detecting, for a bacterial sample, plasmid contigs bins from the  **assembly graph** (provided in <a href="http://gfa-spec.github.io/GFA-spec/">GFA format</a>) generated by the assembler <a href="https://github.com/rrwick/Unicycler">Unicycler</a>. 

The `code` directory contains the source code of PlasBin-flow. 
The `results` directory contains the results described in the paper *PlasBin-flow: A flow-based MILP algorithm for plasmid contigs binning*.

## Installation
PlasBin-flow can be installed from this repository 

~~~
git clone https://github.com/cchauve/PlasBin-flow.git
~~~

PlasBin-flow is written in python (version 3.9.11) and requires the module NetworkX (version 2.7).  

It also requires the ILP solver Gurobi (Version 9.1.2+).  
 
We strongly recommand to run PlasBin-flow using a dedicated python virtual environment (see https://docs.python.org/3.9/library/venv.html).  

A conda package will soon be available.

## Obtaining Gurobi license
PlasBin-flow uses the <a href="https://www.gurobi.com/">Gurobi Solver</a>.
To use PlasBin-flow, a Gurobi license is needed, which is free for academics.  

If you plan to use PlasBin-flow on an institutional High-Performance Computing system (such as a university cluster or a cluster of the <a href="https://alliancecan.ca/en">Digital Research Alliance of Canada</a>), we recommend that you contact a system administrator as it is lilely that such a license is already available.

Alternatively, to be used on a local system or on a local computer (all experiments in the paper *PlasBin-flow: A flow-based MILP algorithm for plasmid contigs binning* have been ran on a laptop computer), you can obtain and install a free academic license following the instructions provided at <a href="https://www.gurobi.com/academia/academic-program-and-licenses/">Gurobi: Always Free for Academics</a>.

## Input
PlasBin-flow requires the following input files:
1. The assembly graph of a bacterial sample, provided in <a href="http://gfa-spec.github.io/GFA-spec/">GFA format</a>, generated by <a href="https://github.com/rrwick/Unicycler">Unicycler</a>. The ability to process GFA files generated by <a href="https://github.com/ncbi/SKESA">SKESA</a> will soon be available.
2. A CSV file describing the mapping of the genes from a plasmid genes database to the contigs of the analyzed assembly graph. The format of this file is the <a href="https://www.ncbi.nlm.nih.gov/books/NBK279684/table/appendices.T.options_common_to_all_blast/">blastn output format 6</a>.
3. A file containing a list of GC content likelihoods (one entry per line)<br/>
Each line of this file contains the probabilities that a particular contig belongs to a plasmid with GC content to various GC content ranges.

## Usage
```
python plasbin_flow.py --ag assembly.gfa --gc gc_probs.csv --map gene_contig_mapping.csv \
		--out output_dir --alpha1 alpha_1 --alpha2 alpha_2 --alpha3 alpha_3 --rmiter rmiter
```
Additional arguments
```
-- out output_dir               Directory where the output files are written
--rmiter			Number of iterations to remove circular components. (default: 50)
--alpha1			Weight of flow term. (default: 1)                              
--alpha2			Weight of GC content term. (default: 1)
--alpha3			Weight of gene density term. (default: 1)
```

## Output

The output is a file `components.csv` where each line represents a plasmid contigs bin in the format `plasmid ID;comma separated list of contigs`.
