# PlasBin-flow: A flow-based MILP algorithm for plasmid contigs binning

## Overview

The identification of plasmids and plasmid genes from sequencing data is an important question regarding antimicrobial resistance spread and other One-Health issues. PasBin-flow is a plasmid bining tool that clusters contigs from a short-read assembly into groups of contigs expected to originate from the same plasmid (called *plasmid contigs bins*). 

PlasBin-flow is based on a Mixed-Integer Linear Programming (MILP) formulation for detecting, for a bacterial sample, plasmid contigs bins from the  **assembly graph** (provided in <a href="http://gfa-spec.github.io/GFA-spec/">GFA format</a>) generated by the assembler <a href="https://github.com/rrwick/Unicycler">Unicycler</a>. 

The `code` directory contains the source code of PlasBin-flow. 

PlasBin-flow is described in the paper *PlasBin-flow: A flow-based MILP algorithm for plasmid contigs binning*.

## Installation
PlasBin-flow can be installed from this repository 

~~~
git clone https://github.com/cchauve/PlasBin-flow.git
~~~

PlasBin-flow is written in python (version 3.9.11) and requires the module NetworkX (version 2.7).  

It also requires the ILP solver Gurobi (Version 9.1.2+).  
 
We strongly recommand to run PlasBin-flow using a dedicated python virtual environment (see https://docs.python.org/3.9/library/venv.html).  

A conda package will soon be available.

## Obtaining Gurobi license
PlasBin-flow uses the <a href="https://www.gurobi.com/">Gurobi Solver</a>.
To use PlasBin-flow, a Gurobi license is needed, which is free for academics.  

If you plan to use PlasBin-flow on an institutional High-Performance Computing system (such as a university cluster or a cluster of the <a href="https://alliancecan.ca/en">Digital Research Alliance of Canada</a>), we recommend that you contact a system administrator as it is lilely that such a license is already available.

Alternatively, to be used on a local system or on a local computer (all experiments in the paper *PlasBin-flow: A flow-based MILP algorithm for plasmid contigs binning* were ran on a laptop computer), you can obtain and install a free academic license following the instructions provided at <a href="https://www.gurobi.com/academia/academic-program-and-licenses/">Gurobi: Always Free for Academics</a>.

## Input
PlasBin-flow requires the following input files:
1. The *assembly graph* of a bacterial sample, provided in <a href="http://gfa-spec.github.io/GFA-spec/">GFA format</a>, generated by <a href="https://github.com/rrwick/Unicycler">Unicycler</a>. The ability to process GFA files generated by <a href="https://github.com/ncbi/SKESA">SKESA</a> will soon be available.
2. A *plasmid gene mapping* TSV file describing the mapping of the genes from a plasmid genes database to the contigs of the analyzed assembly graph. The format of this file is the <a href="https://www.ncbi.nlm.nih.gov/books/NBK279684/table/appendices.T.options_common_to_all_blast/">blastn output format 6</a>.
3. A *GC content* TSV file containing on each line the probability for a contig to originate from a molecule of GC content within a given range (see section 2.5 of the paper *PlasBin-flow: A flow-based MILP algorithm for plasmid contigs binning*).

### Creating the plasmid gene mapping file

### Creating the GC content file
This code helps in generating the *GC content* TSV file mentioned above. It takes as input an assembly graph file in GFA format (same as above) and optionally, a file specifying the endpoints of the GC content intervals. The output generated is a TSV file with one line per contig, with the probablities that a contig originates from a molecule of GC content within the prescribed ranges. 
```
python code/get_gc_probs.py --ag assembly_graph --out output_file
```
where `assembly_graph` is the assembly graph file.

Additional arguments:
```
--gcint			Path to file with GC content interval. This file should contain the endpoints of required GC content intervals (including 0 and 1), one endpoint per line
```

## Usage
```
python code/plasbin_flow.py --ag assembly_graph --gc gc_content_file --map gene_contig_mapping_file \
		--out output_dir --alpha1 alpha_1 --alpha2 alpha_2 --alpha3 alpha_3 --rmiter rmiter
```
where `assembly_graph` is the assembly graph file, `gc_content_file` is the GC content file described above and `gene_contig_mapping_file` is the gene mapping file described above.

Additional arguments:
```
--output_dir        Directory where the output files are written (required)
--output_file       Name of output file (required)
--rmiter			Maximum number of iterations to remove circular components. (optional, default: 50)
--alpha1			Weight of flow term. (optional, default: 1)                              
--alpha2			Weight of GC content term. (optional, default: 1)
--alpha3			Weight of gene density term. (optional, default: 1)
```

## Output
The output of plasbin is a TSV file with each line containing the following information:
```
Plasmid bin			Iteration number associated with plasmid bin
Flow value			Flow value associated with plasmid bin
GC bin				Index of GC content interval associated with plasmid bin
Contigs				List of contigs associated with plasmid bin alongwith their multiplicities in a CSV format
```
Thus a typical line in an output file from plasbin looks as follows:
```
P1	2.5	4	a:2, b:3, c:2, d:1
```
