# PlasBin-flow: A flow-based MILP algorithm for plasmid contigs binning

PlasBin-flow is a tool aimed at extracting from the **assembly graph**
(in <a href="http://gfa-spec.github.io/GFA-spec/">GFA format</a>) of a
bacterial isolate, groups of contigs (called **plasmid bins**), each
groups being assumed to originate from one of a plasmid of the
analyzed isolate.  Currently PlasBin-flow can only process assmbly
graphs genrated by the assembler <a
href="https://github.com/rrwick/Unicycler">Unicycler</a>.

PlasBin-flow is based on a Mixed-Integer Linear Programming (MILP)
formulation that is described in the paper <a
href="https://doi.org/10.1093/bioinformatics/btad250">PlasBin-flow: a
flow-based MILP algorithm for plasmid contigs binning</a>.

## Installation

PlasBin-flow can be installed from this repository
```git clone https://github.com/cchauve/PlasBin-flow.git```

PlasBin-flow is composed of two main scripts:
- [plasbin_flow.py](code/plasbin_flow.py) is the main script used to
  process an assembly graph.
- [plasbin_utils.py](code/plasbin_utils.py) is a script used to tune
  palsbin-flow parameters given a reference dataset and preprocess
  a set of samples prior to processing them with plasbin-flow.

### Prerequisites
PlasBin-flow is written in python (version 3.9.11+);
- [plasbin_flow.py](code/plasbin_flow.py) requires
  - the python module <a href="https://networkx.org/">networkX (version 2.7+)<a/>,
  - the ILP solver <a href="https://www.gurobi.com/">gurobi (version 9.1.2+)</a>.
- [plasbin_utils.py](code/plasbin_utils.py) requires the python modules
  - <a href="https://pandas.pydata.org/">pandas (version 2.0.0+)</a>,
  - <a href="https://matplotlib.org/">matplotlib (version 3.7.0+)</a>,
  - <a href="https://seaborn.pydata.org/">seaborn (version 0.12.2+)</a>,
  - <a href="https://biopython.org/">Biopython (version 1.81+)</a>,
  - <a href="https://scipy.org/">scipy (version 1.10.1+)</a>,
  - <a href="https://numpy.org/">numpy (version 1.24.2+)</a>.

[plasbin_utils.py](code/plasbin_utils.py) also requires <a href="https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html">BLAST+
v2.6.0</a> (commands makeblastdb and blastn).

We strongly recommand to run PlasBin-flow using a dedicated python virtual environment (see https://docs.python.org/3.9/library/venv.html).  

### Obtaining Gurobi license
PlasBin-flow uses the <a href="https://www.gurobi.com/">Gurobi Solver</a>.
To use PlasBin-flow, a Gurobi license is needed, which is free for academics.  

If you plan to use PlasBin-flow on an institutional High-Performance
Computing system (such as a university cluster or a cluster of the <a
href="https://alliancecan.ca/en">Digital Research Alliance of
Canada</a>), we recommend that you contact a system administrator as
it is lilely that such a license is already available.

Alternatively, to be used on a local system or on a local computer
(all experiments in the paper *PlasBin-flow: A flow-based MILP
algorithm for plasmid contigs binning* were ran on a laptop computer),
you can obtain and install a free academic license following the
instructions provided at <a
href="https://www.gurobi.com/academia/academic-program-and-licenses/">Gurobi:
Always Free for Academics</a>.

## Running PlasBin-flow

**Question**: should we take as input a plasmidness file (contig<TAB>plasmidness score) instead of the mappings file?

### Input files

PlasBin-flow requires the following input files (described in details
below) to process a bacterial isolate:
- assembly graph file,
- reference plasmid genes to assembly contigs mapping file;
- GC content probabilities file.

It also requires two numerical parameters defining the *seed contigs*,
that are contigs that are more likely to belong to a plasmid (see
section 2.4 of the PlasBin-flow paper).

#### Assembly graph

The *assembly graph* of a bacterial sample must be the provided in
<a
href="http://gfa-spec.github.io/GFA-spec/">GFA format</a> file
generated by <a
href="https://github.com/rrwick/Unicycler">Unicycler</a> (file
`assembly.gfa`).  We are currently working to add the possibility to
process GFA assembly graphs generated by other assemblers.

**QUESTION**: gzipped or not; plasbin_utils assumes it is. It should
  be uniformized between the two commands.

#### Mapping file

The *plasmid gene mapping* file is a <a
href="https://www.ncbi.nlm.nih.gov/books/NBK279684/table/appendices.T.options_common_to_all_blast/">blastn
output format 6</a> file obtained by mapping a reference set of
plasmid genes to the assembly graph contigs.

This file is used to define an a priori measure of *plasmidness* of
the assembly contigs, defined as the percentage of the contig covered
by mappings (called the *gene density*).

A user of PlasBin-flow can use a custom reference plasmid genes set
(that can be built using the tuning commands of
[plasbin_utils.py](code/plasbin_utils.py) for example) or use the
default database which is provided in
[genes.fasta](database/genes.fasta), based on the plasmids whose NCBI
accession numbers are listed in
[accessions.txt](database/accessions.txt).  A reference plasmid gene
database is simply a FASTA file, with each entry being assumed to be a
gene.

Given a reference database and an assembly graph file, the mapping
file can be generated by the preprocessing commands of
[plasbin_utils.py](code/plasbin_utils.py).

#### GC content probabilities file

As described in section 2.5 of the PlasBin-flow paper, GC content
plays an important tole in the PlasBin-flow optimization model.
PlasBin-flow relies on a set of relative GC content intervals (a
partition of [0,1] into intervals, each defining a range of relative
GC content, called a GC interval) and requires that each contig *c* is
given, for each GC interval *I* a *probability* that *c* originates
from a molecule of relative GC content in the range defined by *I*.  

By default the considered GC intervals of PlasBin-flow are *[0,0.4],
[0.4,0.45], [0.45,0.5], [0.5,0.55], [0.55,0.6], [0.6,1]* but customs
intervals can be computed using the tuning commands of
[plasbin_utils.py](code/plasbin_utils.py) for example).

Given a GC intervals file and an assembly graph, the GC content
probabilities file can be generated by the preprocessing commands of
[plasbin_utils.py](code/plasbin_utils.py).

#### Seeds parameters

The seed parameters are the minimum length *l* and the minimum gene
density *gd*: any contig of length at least *l* and gene density at
least *gd* is a seed contig.

**QUESTION**: default values?

### Usage

**TO DO**: rename options `outdir` as `out_dir` and `outfile` as `out_file`.

To process a sample, given the input files described above, the
command [plasbin_flow.py](code/plasbin_flow.py) can be used as
follows:

```
python code/plasbin_flow.py \
       -ag assembly_graph \
       -gc gc_content_file \
       -map mapping_file \
       -lthr length_threshold \
       -pthr gene_density_threshold \
	-outdir output_dir \
	-outfile output_file
```
where
- `assembly_graph` is the GFA format assembly graph file,
- `gc_content_file` is the GC content probability file described above
- `mapping_file` is the reference plasmid genes to contigs mapping file described above,
- `lthr` and `pthr` are the length and gene density thresholds defining seeds,
- `output_dir` and `output_file` are the directory and file name for the output file.

The output of PlasBin-flow is a TSV file with each line containing the following information:
```
Plasmid bin			Number (ID) associated with the plasmid bin
Flow value			Flow value associated with the plasmid bin (used as a proxy for the plasmid bin copy number)
GC bin				Index of GC content interval associated with the plasmid bin
Contigs				Comma-separated list of contigs associated with plasmid bin along with their multiplicities
```
Thus a line in an output file from PlasBin-flow is as follows:
```
#Pls_ID	Flow	GC_bin		Contigs
P1	2.5	0.4-0.45	a:2,b:3,c:2,d:1
```

### Preprocessing

The script [plasbin_utils.py](code/plasbin_utils.py) allows to
preprocess a set of samples to create the mapping and GC content
probability files for all samples.

The command to create the mapping files for a set of samples is
```
python plasbin_utils.py map_genes_to_ctgs \
       --input_file input_file \
       --out_dir out_dir \
       --tmp_dir tmp_dir \
       --db_file pls_db_file \
       [--out_file out_file]
```
where

- `input_file` is a CSV file with a header containing at least the
  fields `sample` and `gfa` where the column `sample` contains the
  sample name and the column `gfa` the paths to the samples gzipped
  GFA files (one sample per line);
- `out_dir` is the directory where the output of the mapping is
  written to a file named `<sample name>.genes_mappings.txt`.
- `tmp_dir` is directory created to hold temporary files.
- `pls_db_file` is the path to reference plasmid genes database file.
- `out_file` is an optional parameter; if provided, a new CSV file
  `out_file` is created from `input_file` by adding a column
  `genes2ctgs_mappings` that contains the path to the file `<sample
  name>.genes_mappings.txt`.

**TODO** If we use plasmidness instead of mappings as input, add the `gene_density` command.

The command to create the GC content probabilities files for a set of samples is
```
python plasbin_utils.py gc_probabilities \
       --input_file input_file \
       --out_dir out_dir \
       --tmp_dir tmp_dir \
       --gc_intervals gc_intervals_file \
       [--out_file out_file]
```
where the parameters `input_file`, `out_dir`, `tmp_dir` are as above
and `gc_intervals_file` is a file that describes the GC intervals,
with one interval boundary per line (so its first line must be 0, its
last line must be 1 and in between the values must be increasing, an
example is provided in
[gc_intervals.txt](example/input/gc_intervals.txt).

The file created for a sample that contains the GC content
probabilities is names `<sample name>.gc.tsv`.

Here too `out_file` is an optional parameter; if provided, a new CSV
file `out_file` is created from `input_file` by adding a column
`gc_probabilities` that contains the path to the file `<sample
name>.gc.tsv`.

Finally, both preprocessing commands can be ran at once using
```
python plasbin_utils.py preprocessing \
       --input_file input_file \
       --out_dir out_dir \
       --tmp_dir tmp_dir \
       --db_file pls_db_file \       
       --gc_intervals gc_intervals_file \
       [--out_file out_file]
```

**TODO** If we use mappings as input, the code needs to be modified to
  not add gene density files in the output file.

## Tuning PlasBin-flow parameters

PlasBin-flow requires auxiliary data files and parameters in order to process samples:
- reference plasmid genes database used to compute the mapping file and the gene density,
- GC content intervals used to compute the GC content probabilities file,
- parameters defining seeds contigs.

PlasBin-flow comes with default files/values for these (see the
PlasBin-flow paper for a description of how these were obtained):
- reference plasmid genes database: [genes.fasta](database/genes.fasta)
- GC content intervals:     [gc_intervals.txt](example/input/gc_intervals.txt)
- seed contigs parameters:  **QUESTION**: default values

Alternatively, given a *reference set* of samples for which the true
plasmids are known, the script
[plasbin_utils.py](code/plasbin_utils.py) can be used to compute these
parameters and auxiliary files.

### Creating a plasmid gene database
The command
```
python plasbin_utils.py pls_genes_db \
       --input_file input_file \
       --out_dir out_dir \
       --tmp_dir tmp_dir
```
creates a reference plasmid genes database `out_dir/pls.genes.fast`. 

The `file input_file` is a CSV file containing at least the following
fields
- `sample`: sample name,
- `gfa`: path to the gzipped GFA file of the sample,
- `pls_fasta`: path to a gzipped FASTA file containing the sample
  plasmids, each entry being named by the GenBank or RefSeq accession
  number of the plasmid.

The directory `tmp_dir` is used to store temporary files and the
resulting plasmid genes database is the FASTA file
`out_dir/pls.genes.fasta`.

### Computing GC content intervals

The command
```
python plasbin_utils.py gc_intervals \
       --input_file input_file \
       --out_dir out_dir \
       --tmp_dir tmp_dir \
       [--n_gcints n_gcints]
```
computes a GC content intervals file `out_dir/gc.txt`.

The file `input_file` is a CSV file containing at least the following
fields
- `sample`: sample name,
- `pls_fasta`: path to a gzipped FASTA file containing the sample
  plasmids, one per FASTA entry,
- `chr_fasta`: path to a gzipped FASTA file containing the sample
  chromosomes, one per FASTA entry.

The optional parameter `n_gcints` (default value 6) is the number of intervals.

The command generates three files:
- `out_dir/gc.csv`: details of the GC content of all plasmids and chromosomes;
- `out_dir/gc.png`: violin plot of `out_dir/gc.csv`
- `out_dir/gc.txt`: GC content `n_gcints` intervals boundaries,
  obtained by splitting into equal range intervals the GC content
  observed in the plasmids of the reference samples.

We encourage a user too look at the provided files `out_dir/gc.csv`
and `out_dir/gc.png` to assess the GC intervals file and modify it if
deemed relevant.

### Computing ground truth

For a set of samples for which true plasmids are known, the command
```
python plasbin_utils.py ground_truth \
       --input_file input_file \
       --out_dir out_dir \
       --tmp_dir tmp_dir \
       [--out_file out_file] \
       [--pid_threshold p] \
       [--cov_threshold c]
```
computes, for each sample, a TSV file `out_dir/<sample
name>.ground_truth.tsv` that records true plasmid bins. True plasmid
bins for a sample are defined by mapping the sample contigs to the
true plasmids, and discarding any mapping with identity below
`pid_threshold p` (default value 0.95) and contig coverage below
`--cov_threshold c` (default value 0.8).

The file `input_file` is a CSV file containing at least the following
fields
- `sample`: sample name,
- `gfa`: path to the gzipped GFA file of the sample,
- `pls_fasta`: path to a gzipped FASTA file containing the sample
  plasmids, one per FASTA entry.


The resulting ground truth file `out_dir/<sample
name>.ground_truth.tsv` lists contigs mapped to plasmids associated
with the sample. A line in the ground truth TSV file should contain
the name or id of the plasmid in the first column and a contig that
has been mapped to the plasmid in the second column. The file can
contain other information as long as the first two columns contain the
plasmid and contig ids respectively.
```
#Pls_ID Ctg_ID
P1 C1
P1 C2
P2 C3
...
```

If the optional parameter `out_file` is provided it contains the path
to a new CSV file obtained from `input_file` augmented by a column
`ground_truth` that contains the path to the ground truth file
for each sample.

### Computing seed contigs parameters

PlasBin-flow takes two parameters as input as the contig length and
plasmidness threshold to decide which contig from a given sample can
be designated as seeds. Each plasmid bin output by PlasBin-flow should
contain at least one seed contig. Seed parameters are decided by
observing the length and plasmidness of contigs from the set of
reference samples, based on their mappings to true plasmids.

The command
```
python plasbin_utils.py seeds \
       --input_file input_file \
       --out_dir out_dir \
       --tmp_dir tmp_dir \
       --db_file pls_db_file
```
computes a file `out_dir/seeds.txt` that contains the optimal seed
contigs parameters.

Each line of the `out_dir/seeds.txt` file contains a pair
(plasmidness, length) of thesholds separated by a tab. Each pair is
considered as one of the optimal seed contigs parameters for samples
in the reference set.

The file `input_file` is a CSV file containing at least the following
fields
- `sample`: sample name,
- `gfa`: path to the gzipped GFA file of the sample,
- `ground_truth`: path to the ground truth file for the sample in
  format as described above.

The file `pls_db_file` is reference plasmid genes database file.

### Tuning PlasBin-flow

Finally, all parameters tuning steps (but computing ground truth
files) can be done at once with the command
```
python plasbin_utils.py tuning \
       --input_file input_file \
       --out_dir out_dir \
       --tmp_dir tmp_dir \
       [--out_file out_file] \
       [--pid_threshold p] \
       [--cov_threshold c] \
       [--n_gcints n_gcints]
``` 
where `input_file` is a CSV file containing at least the following
fields
- `sample`: sample name,
- `gfa`: path to the gzipped GFA file of the sample,
- `pls_fasta`: path to a gzipped FASTA file containing the sample
  plasmids, one per FASTA entry,
- `chr_fasta`: path to a gzipped FASTA file containing the sample
  chromosomes, one per FASTA entry.

It creates the files `out_dir/gc.[txt,png,csv]`,
`out_dir/gc_intervals.txt`, `out_dir/seeds.txt`,
`out_dir/pls.genes.fasta`, and for each sample `out_dir/<sample
name>.genes_mappings.txt` and `out_dir/<sample
name>.ground_truth.tsv`.

If the optional paramater `out_file` is provided it contains the path
to a new CSV file obtained from `input_file` augmented by a column
`ground_truth` that contains the path to the ground truth file
for each sample.

